# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [1]:
import warnings
warnings.simplefilter('ignore')
import pandas as pd

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [2]:
import sqlite3
con = sqlite3.connect('Great_Outdoors_Data_SQLite/go_sales.sqlite')

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [78]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
tables = pd.read_sql_query(sql_query,con)
product = pd.read_sql_query("SELECT * FROM product",con)
country = pd.read_sql_query("SELECT * FROM country",con)
retailer = pd.read_sql_query("SELECT * FROM retailer_site",con)
salesbranch = pd.read_sql_query("SELECT * FROM sales_branch",con)
returned = pd.read_sql_query("SELECT * FROM returned_item",con)
order = pd.read_sql_query("SELECT * FROM order_details",con)
staff = pd.read_sql_query("SELECT * FROM sales_staff",con)

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [11]:
product['PRODUCTION_COST'] = pd.to_numeric(product['PRODUCTION_COST'], errors='coerce')
filtered_products = product.loc[(product['PRODUCTION_COST'] > 50) & (product['PRODUCTION_COST'] < 100), ['PRODUCT_NAME', 'PRODUCTION_COST']]


print(filtered_products)

                       PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35           

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [13]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'], errors='coerce')

filtered_products = product.loc[(product['MARGIN'] < .2) | (product['MARGIN'] > 0.6), ['PRODUCT_NAME', 'PRODUCTION_COST']]


print(filtered_products)

                     PRODUCT_NAME  PRODUCTION_COST
29                 Hibernator Pad            29.17
32   Canyon Mule Climber Backpack            62.50
100             BugShield Natural             2.00
101               BugShield Spray             2.00
102         BugShield Lotion Lite             2.00
103              BugShield Lotion             2.50
105             BugShield Extreme             2.50


Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [20]:
filtered_country = country.loc[(country['CURRENCY_NAME'] == 'francs') , 'COUNTRY']


print(filtered_country)

0          France
7         Belgium
19    Switzerland
Name: COUNTRY, dtype: object


Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [23]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'], errors='coerce')

# Filter de producten met een marge groter dan 0.5
product_with_high_margin = product[product['MARGIN'] > 0.5]

# Selecteer de introductiedatums
introductiedatums = product_with_high_margin['INTRODUCTION_DATE']

# Creëer een DataFrame met de introductiedatums
introductiedatums_df = pd.DataFrame(introductiedatums, columns=['INTRODUCTION_DATE'])

# Print het resultaat
print(introductiedatums_df)

    INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
17         15-12-2019
79         26-10-2016
100         15-2-2011
101         15-2-2011
102         15-2-2011
103         15-2-2011
105         15-2-2011
106         15-2-2011
107         15-2-2011
108         15-2-2011
109         15-2-2011
114         15-2-2011


Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [39]:
filtered_df = salesbranch.loc[(salesbranch['ADDRESS2'].notna()) & (salesbranch['REGION'].notna()), ['ADDRESS1', 'CITY']]
print(filtered_df)


                            ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam


Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [36]:
filtered_country = country.loc[(country['CURRENCY_NAME'] == 'dollars') | (country['CURRENCY_NAME'] == 'new dollar'), 'COUNTRY']



print(filtered_country)

3            Taiwan
6         Australia
14    United States
15           Canada
Name: COUNTRY, dtype: object


Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [41]:
filtered_df = retailer.loc[(retailer['POSTAL_ZONE'].str.startswith('D')) & (retailer['ADDRESS2'].notna()), ['ADDRESS1', 'ADDRESS2', 'CITY']]
print(filtered_df)

             ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg


## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [52]:
returned['RETURN_QUANTITY'] = returned['RETURN_QUANTITY'].astype(int)

total_returned_products = returned['RETURN_QUANTITY'].sum()
print(total_returned_products)


14306


Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [55]:
total_branch_locations = salesbranch['REGION'].count()
print(total_branch_locations)

15


Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [59]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'], errors='coerce')

laagste_marge = product['MARGIN'].min()
hoogste_marge = product['MARGIN'].max()
gemiddelde_marge = product['MARGIN'].mean().round(2)

print("Laagste marge:", laagste_marge)
print("Hoogste marge:", hoogste_marge)
print("Gemiddelde marge:", gemiddelde_marge)


Laagste marge: 0.17
Hoogste marge: 0.7
Gemiddelde marge: 0.4


Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [69]:
aantal_onbekend_adres2 = retailer[retailer['ADDRESS2'].isnull() | retailer['ADDRESS2'] == ''].count().sum()

print(aantal_onbekend_adres2)

0


Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [77]:
order['UNIT_SALE_PRICE'] = order['UNIT_SALE_PRICE'].astype(float)
order['UNIT_PRICE'] = order['UNIT_PRICE'].astype(float)

discounted_products = order[order['UNIT_SALE_PRICE'] < order['UNIT_PRICE']]
average_discounted_price = discounted_products['UNIT_PRICE'].mean()

print(average_discounted_price)

202.23407425232037


Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [82]:
employee_counts = staff.groupby('POSITION_EN').size().reset_index(name='Aantal')

print(employee_counts)

                    POSITION_EN  Aantal
0                Branch Manager      19
1              District Manager       4
2               General Manager       1
3  Level 1 Sales Representative      13
4  Level 2 Sales Representative      39
5  Level 3 Sales Representative      24
6              Regional Manager       2


Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [90]:
employee_counts_per_phone = staff.groupby('WORK_PHONE').size().reset_index(name='count')

filtered_employee_counts = employee_counts_per_phone[employee_counts_per_phone['count'] > 4]

limited_employee_counts = filtered_employee_counts.head(10)

print(limited_employee_counts)

             WORK_PHONE  count
4     +(44) 121 3505267      5
10  +(61) 03 2982  4242      5
11  +31 (0)20 692 93 94      5
13    +33 1 68 94 52 20      5
18     1 (206) 292-0012      5
19     1 (305) 557-4810      6
20     1 (310) 281-5722      5
21     1 (403) 232-5986      5
22     1 (416) 493-5595      5
23     1 (617) 268-6754      6


## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [1]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [2]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 